### PR TITLE
Improve age computation formula and Variables tutorial comments

### PR DESCRIPTION
**Pull request by MattiSG**
* Tax and benefit system evolution.
* Impacted periods: all.
* Impacted areas: `demographics.age`
* Details:
  - Improve reliability and accuracy of `age` formula
  - Improve variables comments

- - - -

Comments improvements are only partly a knight: PRs such as [ServiceInnovationLab/openfisca-aotearoa#15](https://github.com/ServiceInnovationLab/openfisca-aotearoa/pull/15/files#diff-e97a4b5c683152aba1c810f24751a129R12) show that reusers will copy and paste comments and that can lead to review friction.

### CODE REVIEW

**Comment by MattiSG** [comment]
@fpagnoux CircleCI's flake8 complaint seems a bit strange to me, since it is related to the age formula you wrote would you mind handling it? 🙂 

> ./openfisca_country_template/variables/demographics.py:12:1: F401 'numpy.datetime64' imported but unused

**Comment by benjello** [comment]
Surfing on this PR: @guillett @fpagnoux : should we transpose this to france ?

File: openfisca_country_template/variables/benefits.py [hunk 1]
```diff
@@ -1,12 +1,12 @@
 # -*- coding: utf-8 -*-
 
-# This file defines the variables of our legislation.
-# A variable is property of a person, or an entity (e.g. a household).
+# This file defines variables for the modelled legislation.
+# A variable is a property of an Entity such as a Person, a Family…
```
Comment (fpagnoux) [inline openfisca_country_template/variables/benefits.py]: Shouldn't we use `Household`? There is no `Family` in this model.

File: openfisca_country_template/variables/demographics.py [hunk 2]
```diff
@@ -1,17 +1,26 @@
...
 
-# Import from openfisca-core the common python objects used to code the legislation in OpenFisca
+# Import from openfisca-core the common Python objects used to code the legislation in OpenFisca
 from openfisca_core.model_api import *
-# Import the entities specifically defined for this tax and benefit system
+# Import the Entities specifically defined for this tax and benefit system
 from openfisca_country_template.entities import *
-
+# Import additional NumPy helpers to work on specific data types
 from numpy import datetime64
```
Comment (fpagnoux) [inline openfisca_country_template/variables/demographics.py]: I don't think we are actually using this import anymore, and that's why Flake is complaning.

File: openfisca_country_template/variables/demographics.py [hunk 3]
```diff
@@ -21,14 +30,10 @@ class age(Variable):
     # A person's age is computed according to its birth date.
     def formula(person, period, parameters):
         birth = person('birth', period)
-        return (datetime64(period.date) - birth).astype('timedelta64[Y]')
+        birth_year = birth.astype('datetime64[Y]').astype(int) + 1970
+        birth_month = birth.astype('datetime64[M]').astype(int) % 12 + 1
+        birth_day = (birth - birth.astype('datetime64[M]') + 1).astype(int)
```
Comment (fpagnoux) [inline openfisca_country_template/variables/demographics.py]: > Frankly ugly, but we don't have better yet

**Comment by fpagnoux** [comment]
> Surfing on this PR: @guillett @fpagnoux : should we transpose this to france ?

Probably a good idea, I think there was an issue opened about it somewhere.

File: openfisca_country_template/variables/demographics.py (hunk 2 above)
Comment (MattiSG) [reply openfisca_country_template/variables/demographics.py]: I agree, but I wanted a double check from the person who wrote the initial code :wink:

File: openfisca_country_template/variables/demographics.py (hunk 3 above)
Comment (MattiSG) [reply openfisca_country_template/variables/demographics.py]: I'd _really_ like that a time helper library is researched and vetted for.

**Comment by MattiSG** [comment]
@fpagnoux the tests seem to show there is an issue with the formula 😕

We should probably add more tests to this in any case, including for leap years.

File: openfisca_country_template/variables/demographics.py (hunk 3 above)
Comment (fpagnoux) [reply openfisca_country_template/variables/demographics.py]: We started to ask around, and there are candidates (Maya, Arrow, etc.). Though in this specific case, we need to handle _vectors_ of dates, which will be a clear limitation.
If we were to use a library, we should also probably also use it to handle periods. So IMO this is related to https://github.com/openfisca/openfisca-core/issues/670.

**Comment by fpagnoux** [comment]
> @fpagnoux the tests seem to show there is an issue with the formula 😕

Indeed. I fixed a formula and added a test for leap years. Feel free to add the tests you'll find relevant 🙂.

**Comment by MattiSG** [comment]
That was an annoying rebase, hopefully I got it right :wink:

**Comment by MattiSG** [comment]
> hopefully I got it right

(just kidding, I double-checked that the diff from the resulting rebase to the previous branch was equal to the changes made to `master`)

**Pull request merged**